File: identifySongNotes.m
function identifySongNotes()
    % Plot Settings
    alw = 0.75;    % AxesLineWidth
    fsz = 22;      % Fontsize
    lw = 1.1;      % LineWidth
    msz = 8;       % MarkerSize
    
    % Algorithm Settings
    MIN_NOTE_LEN = 0.12;
    CUTOFF_FREQ = 30;
    DIFF_TOL = 1e-4;
    MAG_TOL = 600;
    MIN_FREQ_SPACING = 50;
    SONG_FILE = 'majorScaleSingle.m4a';
    
    % Load song, discard stereo if present
    [song,Fs] = audioread(SONG_FILE);
    song = song(:,1);
    t = (1:numel(song))/Fs;
    S2IDX = numel(t)/t(end);
    minNoteIdx = round(MIN_NOTE_LEN*S2IDX);
    
    % Filter out higher frequencies
    songFftFilt = fftFilt(abs(song),Fs,CUTOFF_FREQ);
    
    % Get rising edges of signal
    idx = risingEdges(songFftFilt,DIFF_TOL,minNoteIdx);
    
    % Plot of rising edge locations
    fftPlot = figure;
    set(gca, 'FontSize', fsz, 'LineWidth', alw); %<- Set properties
    figure(fftPlot);
    plot(t(idx),songFftFilt(idx),'*r',t,songFftFilt,'-k','linewidth',lw);
    legend('Derivative of Envelope');
    xlabel('Time (s)');
    ylabel('Amplitude');
    title(['Cut-off Frequency = ' num2str(CUTOFF_FREQ) ' Hz']);
    pause;
    
    for ii = 1:numel(idx)
        [freq, mag] = fftMag(song((idx(ii)-minNoteIdx):(idx(ii)+minNoteIdx)),Fs);
        noteIdx = 1:ceil(numel(freq)/2);
        freq = freq(noteIdx);
        mag = mag(noteIdx);
        
        pkIdxs = peakIdxs(mag,MAG_TOL,MIN_FREQ_SPACING);
        
        plot(freq(pkIdxs),mag(pkIdxs),'*r',freq,mag,'-k');
        pause
    end
end

% Find peaks of x above tol atleast spacing apart (need something better
% for dealing with dups)
function idx = peakIdxs(x,tol,spacing)
    % x vector shifted forward and backward one index
    xn = [x(2:end);0];
    xp = [0;x(1:end-1)];
    
    % Peaks above tol
    idx = find(x >= xp & x >= xn & x >= tol);
    
    % Not far enough apart, likely duplicate peaks
    dups = find(diff(idx) < spacing);
    
    dupIdx = [];
    % Remove lower peak
    for jj=1:numel(dups)
        if x(idx(dups(jj))) >= x(idx(dups(jj)+1))
            dupIdx = [dupIdx;dups(jj)+1];
        else
            dupIdx = [dupIdx;dups(jj)];
        end
    end
    
    idx(dupIdx)=[];
end

function idx = risingEdges(x,tol,spacing)
    % Diff and diff shifted forward one
    dx = diff(x);
    dxp = [0;dx(1:end-1)];
    
    % Idx of rising edges
    idx = find(dx >= tol & dxp < tol);
    
    % Removes duplicates
    dupIdx = find(diff(idx) < spacing)+1;
    idx(dupIdx) = [];
end

File: Python/identifySongNotes.py
import numpy as np
from scipy import signal
from time import time

def identifySongNotes(songChunk, Fs, filter_b, filter_a, note_detected, note_time):

    # Algorithm Settings
    MIN_NOTE_LEN = 0.12
    MIN_NOTE_IDX = MIN_NOTE_LEN * Fs
    DIFF_TOL = 4.8

    # Filter out higher frequencies
    #FILTER_CUTOFF = 30.0
    #FILTER_ORD = 4
    #filter_b, filter_a = signal.bessel(FILTER_ORD, [FILTER_CUTOFF / (Fs / 2.0)], btype='low', analog=False) # Do this outside loop
    songChunkFilt = signal.filtfilt(filter_b, filter_a, np.abs(songChunk)) # type: np.ndarray

    note_freq = 0
    #while True:
    # If no note has been detected, search for one
    if not note_detected:
        note_idx = np.nonzero(np.diff(songChunkFilt) >= DIFF_TOL)[0] # type: np.ndarray
        if note_idx.size >= 1:
            note_detected = True

            # Get first note index
            note_idx = note_idx[0]
            note_time = time()

            # Get dominant frequency of note
            fftSongChunk = np.fft.fft(songChunkFilt)
            absFftSongChunk = np.abs(fftSongChunk)
            freqSongChunk = np.fft.fftfreq(fftSongChunk.size, 1.0/Fs)

            note_freq = freqSongChunk[np.argmax(absFftSongChunk - np.mean(absFftSongChunk))]

    # Reset the flag after enough time has passed
    elif time() - note_time >= MIN_NOTE_LEN:
        note_detected = False

    return {'note_detected': note_detected, 'note_time': note_time, 'note_freq': note_freq}


# UNUSED FOR NOW
# Find peaks of x above tol atleast spacing apart(need something better
# for dealing with dups)
def peakIdxs(x, tol, spacing):
    # x vector shifted forward and backward one index
    xn = np.c_[x[2:-1],0]
    xp = np.c_[0,x[:-2]]

    # Peaks above tol
    idx = np.nonzero(np.logical_and(np.logical_and(x >= xp, x >= xn),x >= tol))[0]

    # Not far enough apart, likely duplicate peaks
    dups = np.nonzero(np.diff(idx) < spacing)[0]

    dupIdx = np.array(0)
    # Remove lower peak
    for jj in range(dups):
        if x[idx[dups[jj]]] >= x[idx[dups[jj] + 1]]:
            dupIdx = np.c_[dupIdx, dups(jj) + 1]
        else:
            dupIdx = np.c_[dupIdx, dups(jj)]
    dupIdx = np.delete(dupIdx,0)
    idx = np.delete(idx,dupIdx)
    return idx


def detectRisingEdge(x, Fs, tol, spacing):
    # Diff and diff shifted forward one
    dx = np.diff(x)*Fs
    dxp = np.c_[0,dx[:-2]]

    # Idx of rising edges
    #idx = np.nonzero(np.logical_and(dx >= tol,dxp < tol))[0]
    idx = np.nonzero(dx >= tol)[0]

    # Removes duplicates
    #dupIdx = np.nonzero(np.diff(idx) < spacing)[0] + 1
    #idx = np.delete(idx,dupIdx)
    return idx[0]